### PR TITLE
Finished DTOs (Activity and User), Finished PUT Endpoints

### LINKED ISSUE #23: Implement PUT endpoint for the profile update
#12 , #10 and #8 

UserService Implementation
Tests and Adapt Mappers (depends on #66)

### LINKED ISSUE #64: Add Activity DTO
- [ ] Add DTOs (GET/POST/PUT...)
- [ ] Update DTOMapper

related to #18 & #21

### LINKED ISSUE #66: Add User DTOs
- [ ] Add DTOs (GET/POST/PUT...)
- [ ] Update DTOMapper

related to #39

### LINKED ISSUE #74: Adapt email and phone validation
#12 , #10 and #8
Adapt email and phone validation to not have duplicate check.

### LINKED ISSUE #75: Clean up Activity DTOs
related to #18 & #21

Adapt Activity DTOs and Activity DTO Tests

### PR DESCRIPTION
**Pull request by AleksandarRistic99**
closes #64 
closes #23 
closes #66 
closes #75
closes #74 

- Added many tests in UserServiceTest, ControllerTest, UserServiceIntegrationTest, and DTO Mapper Tests
- pleas take a look at the Activity DTOs, to see if they match the requirements 
- added UserGetDTOProfile that gives public data to the frontend (UserGetDTO gives private data)
- UserPutDTO is for all the changes a user want to make on his profile/data
- added activityMapper
- take a look at the GetMapping getUserProfile in the UserController--> my idea is to return a UserGetDTO if a user is on own profile and to return a UserGetDTOProfile if user is on an other users profile
- added email and phone validation to handle invalid email and phone numbers before they reach database

NOTE: adding all the tests enabled a test coverage of 76% on the methods --> lets try to stay on that coverage to not have a stressful milestone 4 :)

image

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/userDTO/UserPutDTO.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package ch.uzh.ifi.hase.soprafs21.rest.dto.userDTO;
+
+import ch.uzh.ifi.hase.soprafs21.constant.Gender;
+import ch.uzh.ifi.hase.soprafs21.entities.UserInterests;
+
+public class UserPutDTO {
+
+    private String email;
```
Comment (SamuelBruegger) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/userDTO/UserPutDTO.java:8]: should the user be able to update his email?

Comment (AleksandarRistic99) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/userDTO/UserPutDTO.java:8]: Good point, i think they should have the possibility for cases like if they change university (uzh --> eth for example) and then to not be forced to register again, they could just change the email. But this is a point we can discuss on monday 👍

**Comment by amoscalamida** [comment]
@AleksandarRistic99 Handle and Rethrow Validation Exceptions

<code>
try {
            // saves the given entity but data is only persisted in the database once flush() is called
            newUser = userRepository.save(newUser);
            userRepository.flush();
        }
        catch (ConstraintViolationException ex) { handleValidationError(ex); 
}

private void handleValidationError(ConstraintViolationException ex) {
        AtomicReference<String> exceptions = new AtomicReference<>();
        exceptions.set("Validation failed:\n");
        ex.getConstraintViolations().forEach(violation -> {
            exceptions.set(exceptions + violation.getMessage() + "\n");
        });
        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, exceptions.toString());
    }
</code>

**Review by SamuelBruegger — APPROVED** [review]

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UserController.java [hunk 2]
```diff
@@ -86,47 +89,69 @@ public void userPing(@PathVariable Long userId, @RequestHeader("Auth-Token")Stri
...
 
-    @GetMapping("/users/{userId}/profile")
-    @ResponseStatus(HttpStatus.OK)
-    public void getUserProfile(@PathVariable Long userId, @RequestHeader("Auth-Token")String token){
+        // if user is on his own profile, he can see additional data like email --> return UserGetDTO
+        try{
+            if(userService.isUserAuthenticated(userId,token)){
+                return DTOMapperUser.INSTANCE.convertEntityToUserGetDTO(userFromRepo);
+            }
+        }catch(ResponseStatusException ignored){}
```
Comment (Taremeh) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UserController.java:132]: Why try & catch?

**Review by Taremeh — APPROVED** [review]

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UserController.java (hunk 2 above)
Comment (Taremeh) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UserController.java:132]: Nevermind, it's b.c. of the auth function.

**Pull request merged**